drivers/rtc/rtc_rv8263.c

```diff
@@ -0,0 +1,918 @@
+        goto rv8263c8_update_enable_timer_exit;
+    }
+
+    // Configure the countdown timer for 1 Hz and enable pulse interrupts.
+    regs = (0x02 << 3) | (0x01 << 2) | (0x01 << 1) | (0x01 << 0);
+    err = i2c_burst_write_dt(&config->i2c_bus, RV8263C8_REGISTER_TIMER_MODE, &regs, sizeof(regs));
+
+rv8263c8_update_enable_timer_exit:
+    k_spin_unlock(&data->lock, key);
+
+    return err;
+}
+
+/**
+ * @brief
+ *
+ * @param p_dev
+ */
+static void rv8263_process_update(const struct device *p_dev)
+{
+    struct rv8263c8_data *data = p_dev->data;
+    const struct rv8263c8_config *config = p_dev->config;
+
+    if (data->update_cb != NULL) {
+        LOG_DBG("Calling update callback");
+        data->update_cb(p_dev, data->update_cb_data);
+    }
+
+    rv8263c8_update_enable_timer(p_dev);
+    i2c_reg_update_byte_dt(&config->i2c_bus, RV8263C8_REGISTER_CONTROL_2, RV8263C8_BM_TF, RV8263C8_BM_TF);
+}
+
+#if defined(CONFIG_RTC_RV8263_UPDATE_OWN_THREAD)
+/**
+ * @brief
+ *
+ * @param p_arg1
+ * @param p_arg2
+ * @param p_arg3
+ */
+static void rv8263c8_update_thread_func(void *p_arg1, void *p_arg2, void *p_arg3)
+{
+    const struct device *dev = p_arg1;
+    struct rv8263c8_data *data = dev->data;
+
+    while (1) {
+        k_sem_take(&data->update_sem, K_FOREVER);
+        rv8263_process_update(dev);
+    }
+}
+#elif defined(CONFIG_RTC_RV8263_UPDATE_GLOBAL_THREAD)
+/**
+ * @brief
+ *
+ * @param p_work
+ */
+static void rv8263c8_update_worker(struct k_work *p_work)
+{
+    struct rv8263c8_data *data = CONTAINER_OF(p_work, struct rv8263c8_data, alarm_work);
+
+    LOG_DBG("Process update interrupt from worker");
+
+    rv8263_process_update(data->dev);
+}
+#endif
+#endif
+
+/**
+ * @brief           Set time function for RTC API.
+ *
+ * @param p_dev     Pointer to device strucure
+ * @param p_timeptr
+ * @return int      0 when successful
+ */
+static int rv8263c8_time_set(const struct device *p_dev, const struct rtc_time *p_timeptr)
+{
+    int err;
+    uint8_t regs[7];
+    struct rv8263c8_data *data = p_dev->data;
+    const struct rv8263c8_config *config = p_dev->config;
+
+    if ((p_dev == NULL) || (p_timeptr == NULL)) {
+        return -EINVAL;
+    }
+
+    LOG_DBG("Set time: year = %u, mon = %u, mday = %u, wday = %u, hour = %u, min = %u, sec = %u",
+            p_timeptr->tm_year, p_timeptr->tm_mon, p_timeptr->tm_mday, p_timeptr->tm_wday, p_timeptr->tm_hour, p_timeptr->tm_min,
+            p_timeptr->tm_sec);
+
+    regs[0] = bin2bcd(p_timeptr->tm_sec) & SECONDS_BITS;
+    regs[1] = bin2bcd(p_timeptr->tm_min);
+    regs[2] = bin2bcd(p_timeptr->tm_hour);
+    regs[3] = bin2bcd(p_timeptr->tm_wday);
+    regs[4] = bin2bcd(p_timeptr->tm_mday);
+    regs[5] = bin2bcd(p_timeptr->tm_mon);
+    regs[6] = bin2bcd((p_timeptr->tm_year % 100));
+
+    k_spinlock_key_t key = k_spin_lock(&data->lock);
+    err = i2c_burst_write_dt(&config->i2c_bus, RV8263C8_REGISTER_SECONDS, regs, sizeof(regs));
+    k_spin_unlock(&data->lock, key);
```

Comment:
Please use a semaphore to lock/unlock here, a spinlock simply disables interrupts while locked, if the i2c device uses interrupts to send/receive data, the burst write will fail as its being performed with the spinlock locked.

use a semaphore like this
```
/* init to behave as binary and initially available semaphore */
struct k_sem lock;
k_sem_init(&lock, 1, 1);

/* lock semlock */
(void)k_sem_take(&lock, K_FOREVER);

/* unlock semlock */
k_sem_give(&lock);
```
The sem will block the thread if it is already taken, while keeping interrupts enabled.

This video explains it really well https://www.youtube.com/watch?v=vla8DTsttyA

Reply:
Thanks, I will change it. But I´m wondering because I took some of the code structure from other drivers. I. e. the spinlock is taken from the MC14618 driver. What is the difference why it can be used in this driver?